Documentation Refactor

hogepodge opened the pull request:
RFC: https://github.com/apache/tvm-rfcs/blob/main/rfcs/0027-formalize-documentation-organization.md
Tracking Issue: #8987

Stage 1 of the documentation refactor reorganizes the docs structure,
moving files (without content changes) and adding new scaffolding to
generate the proper document tree.

It does not address naming, style, content, links, or other existing
content in documents that were moved. State 2 will address fixing these
issues with existing content.

Major changes include but are not limited to:

* Dividing the existing tutorials into two sections:
  * Tutorials
  * How Tos
* Moving all of the existing tutorials out of the `/tutorial`
  directory and into the more general `/gallery` directory.
* Breaking up how-tos into individual sections for more
  flexibility and more consistent rendering.
* Moving content into new classifications:
  * `/docs/arch` for architecture guides
  * `/docs/reference` for API guides and other reference material
  * `/docs/topic` for topic specific guides such as microTVM and VTA
  * Restructuring `/docs/dev`
* Adding a table of contents to the doc index
* Adding instructions on how to install using third-party tlcpack

Stage 2 of the documentation refactor fixes naming and links
in the documentation to be consistent with the overall structure.

Major changes include:

* an update to how to contribute to docs.
* several updated index pages with title changes to match
  the organization style and bring consistency to the sections
* expanded descriptions of some page collections
* fixed links

hogepodge commented:
Addressed missing links and references in documentation.

tqchen commented:
Thanks Chris, the overall structure looks good and thanks for the effort in preserving the contribution history. 
Looks like we still need to update the CI script in order to get the things into a ready state. 

See related folder here
- https://github.com/apache/tvm/blob/main/tests/scripts/task_python_docs.sh#L42
- https://github.com/apache/tvm/blob/main/tests/scripts/task_sphinx_precheck.sh#L30

Likely we need to update these to the new folder locations, and in the case of `ci_logs`, copy them over.

tqchen commented:
BTW, I wanted to acknowledge and thank you for all the effort you have put into the documentation. Having strong user and developer documentation is critical to TVM’s mission to provide unity to the ML Acceleration ecosystem,and the entire community stands to benefit from better documentation. 

The overall direction of documentation changes looks great and layed out clearly in the RFC. Sorry for the final nit-picks but this are the final steps toward closing the loop of our documentation refactor effort. Lets work together to land this and please let us know what we can do to help.

tmoreau89 commented:
Thank you @hogepodge for this colossal PR and equally colossal effort! These changes will have a hugely positive impact on improving TVM's approachability for the community at large. Thanks for moving the needle in the right direction

denise-k commented:
Huge, huge thanks to @hogepodge for spending so much time and energy to make our documentation shine! :sparkles:

hogepodge commented:
> Thanks Chris, the overall structure looks good and thanks for the effort in preserving the contribution history. Looks like we still need to update the CI script in order to get the things to unblock the ci
> 
> See related lines here
> 
> * https://github.com/apache/tvm/blob/main/tests/scripts/task_python_docs.sh#L42
> * https://github.com/apache/tvm/blob/main/tests/scripts/task_sphinx_precheck.sh#L30
> 
> Likely we need to update these to the new folder locations, and in the case of `ci_logs`, update them to match the new location of auto_scheduler(should be `gallery/how_to/tune_with_autoscheduler`?). Please let me know if you need additonal help.

Thanks! I've updated the scripts. A little while ago @Lunderberg changed the build structure to use both a `_staging` and a `_build` directory. I've updated the scripts to see the new locations and to only remove those directories (as the main tree is not being written in to any more).

tqchen commented:
The remaining error seems was due to tvm not being available in some of the how to example executions.  It might be helpful to run the following command to quickly check a specific tutorial to debug

```
TVM_TUTORIAL_EXEC_PATTERN=/path/to/specific_file\.py make html
```

It could be related to the staging setup, the way we setup env variables (cc @Lunderberg to see if you have any ideas).

hogepodge commented:
Thanks for the suggestion on that. Whether I'm building the entire document set, or just the individual how_tos that are failing in CI, I'm unable to reproduce the failures locally.

hogepodge commented:
I just pushed a new change to see if we can reproduce the same failures in CI.

Lunderberg commented:
Hmm.  It looks like it's having trouble finding the `libtvm.so` file, but it finds it correctly during the initial import.  From a quick 10-minute code dive, it looks like `sphinx_gallery` calls [`memory_profiler.memory_usage`](https://github.com/pythonprofilers/memory_profiler/blob/master/memory_profiler.py#L267).  This starts a subprocess, which doesn't inherit the `sys.path` updates made in [our `conf.py`](https://github.com/hogepodge/tvm/blob/docs-refactor-respin/docs/conf.py#L55).

Can you try adding `os.environ['PYTHONPATH'] = os.pathsep.join(sys.path)` to the `conf.py`, just after the modifications to `sys.path`?  The environment variable should be passed to subprocesses, where the value of `sys.path` is not.

hogepodge commented:
@manupa-arm looks like the same issue

hogepodge commented:
Fixed linting error in conf.py

hogepodge commented:
@Lunderberg I just came to the same conclusion, and addressed it by making a `tvm_path_rel` variable to use in the Sphinx Gallery `ExplicitOrder`.

tqchen commented:
Thanks @hogepodge @Lunderberg @areusch . All the comments are resolved. 

I am merging this in given that this is a refactor that could conflict/block future changes and would benefit from prioritization. Let us send followup PRs for future tweaks.
